# 01 Importing Libraries

In [3]:
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [4]:
# creating a reusable path to my data
path = r'C:\Users\Vaiva\Documents\CareerFoundry\Achievement 4 07-2023 Instacart Basket Analysis'

In [5]:
# importing products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
# importing orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03 Consistency Checks with products.csv

## Mixed-Type Data

In [7]:
# create a dataframe
df_test = pd.DataFrame()

In [8]:
# create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# changing the data type to string
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# creating a subset of only missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# checking the size of the data frame
df_prods.shape 

(49693, 5)

In [16]:
# creating a clean producs subset
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [18]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
# create a new dataframe without the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04 Exporting Dataframe

In [23]:
# exporting products_checked.scv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 05 Task

### Step 2 Inconsistencies with orders_wrangled.csv

In [25]:
# Checking for inconsistencies in orders.csv
df_ords.describe().astype(int)

user_id  order_id  order_number  order_day_of_week  order_hour_of_day  \
count  3421083   3421083       3421083            3421083            3421083   
mean    102978   1710542            17                  2                 13   
std      59533    987581            17                  2                  4   
min          1         1             1                  0                  0   
25%      51394    855271             5                  1                 10   
50%     102689   1710542            11                  3                 13   
75%     154385   2565812            23                  5                 16   
max     206209   3421083           100                  6                 23   

       days_since_prior_order  
count                 3214874  
mean                       11  
std                         9  
min                         0  
25%                         4  
50%                         7  
75%                        15  
max                        30

#### No inconsistencies noted

In [ ]:
df_ords.head()

In [ ]:
df_ords.tail()

### Step 3 Checking for mixed type data

In [ ]:
# check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### There are no mixed data types found

In [ ]:
# check which data types each column has
for column in df_ords.columns:
    print(pd.api.types.infer_dtype(df_ords[column]))

### Step 5 Checking for missing values

In [ ]:
# checking for missing values in df_ords
df_ords.isnull().sum()

#### There are 206209 missing values in days_since_prior_order column

In [ ]:
# creating a subset to see the missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [ ]:
df_nan_ords

In [ ]:
# checking if all order_number values are 1, or if there are others
df_nan_ords[df_nan_ords['order_number'] != 1]

#### Based on the data dictionary, all order_numbers with values 1 means that it's the first order and there are no prior orders made, therefore the missing values make sense.

### Step 6 Addressing the missing values

#### I would leave it as it is as these missing values are there on purpose

### Step 7 Check for duplicate values

In [ ]:
# checking for duplicate values in df_ords
df_dups_ords = df_ords[df_ords.duplicated()]

In [ ]:
df_dups_ords

#### There are no duplicate rows in df_ords

### Step 8 How will the duplicates be handled?

#### There are no duplicates, but if there were, I would use df_ords_no_dups = df_ords.drop_duplicates()

### Step 9 Exporting Dataframe

In [ ]:
# Exporting final df_ords without the index
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)